src/protocol/crsf_uart.c

```diff
@@ -276,31 +276,52 @@ static void build_rcdata_pkt()
 static const u8 rxframes[][26];
 #endif //EMULATOR
 
+static enum {
+    ST_DATA1,
+    ST_DATA2,
+} state;
+
+static volatile int mixer_sync;
+static u16 mixer_runtime;
 static u16 serial_cb()
 {
-    build_rcdata_pkt();
-    UART_Send(packet, sizeof packet);
+    switch (state) {
+    case ST_DATA1:
+        CLOCK_RunMixer(&mixer_sync);    // clears mixer_sync, which is then set when mixer update complete
+        state = ST_DATA2;
+        return mixer_runtime;
+
+    case ST_DATA2:
+        if (!mixer_sync && mixer_runtime < 2000) mixer_runtime += 50;
```

Comment:
The algorithm checks if the mixer calculations completed before being used.  If it did not then add 50 microseconds to its next runtime.  The mixer_runtime variable is used in the callback state machine to adjust when the mixer is run.

The forum thread describes the testing - a model with the maximum number of mixes takes about 1500us to run the mixers.  Simple models take much less time. Here mixer_runtime is initialized to 50us.  If that's not enough for the mixers to complete, it is incremented by 50us each packet interval until the mixers do complete.  This keeps the latency at the minimum working value (within 50us or so).  It only takes a few packet intervals at startup time to reach the working value.

Reply:
Then I understand. I can help take a look mixer code to see why it is so slow. can u point me the exact function here?

if mixer_sync is not finish, shall we just return 50 to give us another try in 50us later?

Reply:
A good place to start is MIXER_CalcChannels.

No that would change the protocol timing.  The protocols that dynamically adjust the runtime given to the mixer calculations quickly adjust the runtime up to ensure the mixer calculations are complete.  For a typical 6-channel model the testing I've done shows the mixer runtime is about 450-500 microseconds.  Since the runtime is adjusted in 50 us steps it takes less than a dozen packet intervals to reach a good value - less than 200 ms.

Reply:
got you. I think your logic is basically try to find out the max latency of mixer calculation. Ultimately, we should limit the calculation time of mixer.